shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -414,6 +414,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
                           val fv = freshVar(noProv,
                             S(f._1.name.drop(f._1.name.indexOf('#') + 1)) // strip any "...#" prefix
                           )(1).tap(_.upperBounds ::= f._2.ub)
+                          // ? not sure if we can do this
```

Comment:
This is a bit funky. It may work in most cases but that's not really the logic we want. Here, we should decide based on whether the field was declared mutable or not. I think you should do what I suggested on Discord:

> it may be better to make the LB of FieldType an Option, and if it's None it means it's immutable

Then things will be a lot cleaner, including here.

This can also give better errors.

Reply:
So it means that we're using `None` for place of `BotType` originally(for immutable ones), and use `Some(T)` for the mutable ones? @_@

Reply:
Yes, exactly!

Reply:
But it seems that the lower bound would always equals to the upper bound for mutable fields, so what is the difference between this way and simply using a flag @_@
Or is there scenario where the lower bound is different?

Reply:
The problem is that if we used a single bound, it would be invariant, which would make other parts of the compiler awkward because they currently assume that everything is either covariant or contravariant. For example, merging two mutable tuples in type simplification would require making the two types equal by constraining, but currently no constraining occurs in simplification. It's more of a technical detail, really.